src/lib.rs

```diff
@@ -171,108 +170,146 @@ use std::sync::atomic::{AtomicIsize, Ordering, ATOMIC_ISIZE_INIT};
 
 use apm::Listener;
 use bson::Bson;
-use common::{ReadPreference, ReadMode, WriteConcern};
+use common::{ReadPreference, WriteConcern};
 use connstring::ConnectionString;
 use db::{Database, ThreadedDatabase};
 use error::Error::ResponseError;
 use pool::PooledStream;
-use stream::StreamConnector;
-use topology::{Topology, TopologyDescription, TopologyType, DEFAULT_HEARTBEAT_FREQUENCY_MS,
-DEFAULT_LOCAL_THRESHOLD_MS, DEFAULT_SERVER_SELECTION_TIMEOUT_MS};
+use stream::ConnectMethod;
+use topology::{Topology, TopologyDescription, DEFAULT_HEARTBEAT_FREQUENCY_MS,
+               DEFAULT_LOCAL_THRESHOLD_MS, DEFAULT_SERVER_SELECTION_TIMEOUT_MS};
 use topology::server::Server;
 
 /// Interfaces with a MongoDB server or replica set.
 pub struct ClientInner {
-    /// Indicates how a server should be selected for read operations.
-    pub read_preference: ReadPreference,
-    /// Describes the guarantees provided by MongoDB when reporting the success of a write
-    /// operation.
-    pub write_concern: WriteConcern,
     req_id: Arc<AtomicIsize>,
     topology: Topology,
     listener: Listener,
     log_file: Option<Mutex<File>>,
 }
 
 /// Configuration options for a client.
-#[derive(Default)]
-pub struct ClientOptions {
+pub struct Connector {
```

Comment:
Given that `ClientOptions` is being converted to a builder for clients, I felt that it should be named better. My quick and dirty best approximation for how I view it is `Connector`, but there likely is a better way to describe it (maybe something relating to the r2d2 trait it will implement, `ManageConnection`).